### ISSUE TITLE
Integrate ZTest

### ISSUE THREAD

**Issue by rsmmr** [zeek/zeek#1475]
We are aiming to upstream https://github.com/corelight/ztest to facility unit testing for Zeek scripts.

**Comment by rsmmr** [comment]
@rvictory @corelight-chris You guys had expressed interest in shepherding ZTest into upstream, would be great if we could get it into 4.1.

**Comment by rsmmr** [comment]
Let's see if we can pull this into 4.1 still, @bbannier is going to take a look.

**Comment by bbannier** [comment]
As an example I rewrote `testing/btest/scripts/base/protocols/ssh/set_version.zeek` with ZTest (here: against the version of `SSH::set_version` in zeek-4.0.2).

```.zeek
# @TEST-EXEC: zeek

@load base/bif/event.bif.zeek
@load base/protocols/ssh
@load ztest

module SSH;

# Creates a mock connection. This connection is good enough for e.g.,
# `SSH::set_version`, but not in line with what Zeek considers active
# connections.
function make_conn(server: string, client: string): connection
	{
	local c: connection;
	c$uid = "uid";

	local id: conn_id;
	id$orig_h = 127.0.0.1;
	id$resp_h = 127.0.0.1;
	id$orig_p = 40/tcp;
	id$resp_p = 40/tcp;
	c$id = id;

	local ssh: SSH::Info;
	ssh$ts = network_time();
	ssh$server = server;
	ssh$client = client;
	c$ssh = ssh;

	SSH::set_session(c);

	c$ssh$version = 0;
	return c;
	}

type Weird: record {
	name: string;
	id: conn_id;
	uid: string;
	addl: string;
	source: string;
};

# A list of weirds raised. These are filled by `expired_conn_weird`
# and can be accessed LIFO-style with `last_weird`.
global weirds: vector of Weird;

# While `SSH::set_version` triggers a `conn_weird` we are dealing with mock
# connections which since they are injected are always considered expired by
# Zeek.
event expired_conn_weird(name: string, id: conn_id, uid: string, addl: string, source: string)
	{
	weirds += Weird($name = name, $id = id, $uid = uid, $addl = addl, $source = source);
	}

# Helper function to get retrieve the last weird. The weird is returned and removed from our internal state.
function last_weird(): Weird
	{
	ZTest::assert_greater_than(|weirds|, 0, "no new weird available");

	# NOTE: Since the previous assert didn't abort the remainder of the
	# script/test below element access can produce a segfault if the vector was
	# empty.

	# Pop last element.
	local last = weirds[|weirds| - 1];
	weirds = weirds[0:|weirds| - 2];

	return last;
	}

const v1 = "SSH-1.5-OpenSSH_6.2";
const v199 = "SSH-1.99-OpenSSH_3.1p1";
const v2 = "SSH-2.0-OpenSSH_5.9";

ZTest::test_suite("SSH::set_version");

ZTest::test("SSH1 vs SSH1", function() {
	local c = make_conn(v1, v1);
	SSH::set_version(c, "");
	ZTest::assert_equal(c$ssh$version, 1, "");
});

ZTest::test("SSH199 vs SSH1", function() {
	local c = make_conn(v1, v199);
	SSH::set_version(c, "");
	ZTest::assert_equal(c$ssh$version, 1, "");
});

ZTest::test("SSH2 vs SSH2", function() {
	local c = make_conn(v2, v2);
	SSH::set_version(c, "");
	ZTest::assert_equal(c$ssh$version, 2, "");
});

ZTest::test("unable to extract full server version", function() {
	local c = make_conn("SSH", v1);
	c$ssh$version = 42;
	SSH::set_version(c, "");
	ZTest::assert_equal(c$ssh$version, 0, "");
	# NOTE: we should check against the real weird here.
	ZTest::assert_equal(last_weird(), Weird($name="", $id=c$id, $uid=c$uid, $addl="", $source=""), "");
});

ZTest::run_tests();
```

As of zeek-4.0.2 this produces the following output:

```
zeek set_version.zeek

Running Test Suite: SSH::set_version
------------------------------------
Running Test: SSH199 vs SSH1

Running Test: unable to extract full server version

Assert Failed:  (SSH::set_version/unable to extract full server version)
Expected: 42
Actual: 0
Backtrace:
#0: print_backtrace(show_args: bool = T, one_line: bool = F, one_line_delim: string = |, skip: count = 1, to_file: file of string = file "/dev/stdout" of string) at /Users/bbannier/src/ztest/scripts/ztest.zeek:92
#1: ZTest::print_failure_backtrace() at /Users/bbannier/src/ztest/scripts/ztest.zeek:142
#2: ZTest::mark_result(succeeded: bool = F, expected_string: string = 42, actual_string: string = 0, message: string = ) at /Users/bbannier/src/ztest/scripts/ztest.zeek:165
#3: ZTest::assert_equal(expected: any = 42, actual: any = 0, message: string = ) at ./set_version.zeek:100
#4: lambda_<3146710000385790034>() at /Users/bbannier/src/ztest/scripts/ztest.zeek:286
#5: ZTest::run_tests() at ./set_version.zeek:104

Assert Failed: no new weird available (SSH::set_version/unable to extract full server version)
Expected: 0.0 to be greater than 0.0
Actual: Was less than or equal to
Backtrace:
[2]    65995 segmentation fault  ZEEKPATH=~/src/ztest/scripts:`zeek-config --zeekpath` zeek set_version.zeek
```

This to note:

- in order to use ZTest we always need to explicitly construct test suites and test cases, assertion macros are not available outside of this framework
- test cases are logged nicely
- failed assertions are logged nicely, including backtraces
- not used here: logging level can be controlled
- BTest here used only as a runner, but e.g., not for output diffing. Instead we can keep code producing output and checking it in one place.
- there is currently no real way to check preconditions, execution continues after assertion failures which can lead to weird, cascading errors, e.g., above segfault.

I filed corelight/ztest#2 to collect possible improvements.

**Comment by rsmmr** [comment]
For me, the main conceptual question is if we want to standardize on using the `test-suite`/`test` framing when running through `btest` as the driver. As Benjamin says, the main alternative would be just using `assert` directly inside code, without enclosing it inside `test`.

I remain somewhat undecided, but am leaning towards keeping the structure as it makes the test units explicit and matches other unit testing systems, at the price of some verbosity. However, we could allow skipping `test-suite`  as there'll mostly likely be use a single one in this model anyways.

**Comment by ckreibich** [comment]
I'm a bit unsure about our goals here, so let me share some thoughts about unit testing with Zeek.
- I definitely agree that there's scope for a script-level testing framework, because there isn't one and it would be very handy.
- That said, I'm wondering whether we should step back a bit and figure out what it would take to really nail unit testing in Zeek. I believe we didn't have that brainstorming session before ZTest appeared. To me the ability to unit-test well isn't just that I can run assertions on groups of functions, though that clearly helps. It's that I need to be able to arrange the surrounding code so that I can meaningfully control the unit under test. Usually this means I need to mock out some surrounding context, call things flexibly, verify side effects, invoke things so testing happens selectively, etc. It is not clear to me right now where Zeek is in that regard.
- On that note, with Zeek being a distributed event based system, it gives me pause how the ZTest docs tell you that in order to deal with Zeek's event-driven nature you should move code out of the event handlers and into functions. Sure, it's what ZTest knows well. But it's not very aligned with script reality, and it can (at least for now) imply performance overhead if you do this pervasively. If we want to properly nail unit testing in Zeek, we should think about how this could fully embrace eventing. In this context I'm excited about the low-level ability to serialize Zeek events to JSON (via Dominik's current CAF work, I believe), because it would give us a way to squirrel away events (and potentially edit them in a text editor, modify them via tooling, etc) and use those for driving unit tests tests.
- Similarly, logs are central to Zeek, so it'd be good to think about what a unit testing framework could provide here. For example, is it feasible/desirable to have something like `ZTest::assert_log_entry(...)`, a bit similarly to how unit testing in other languages lets you verify that an exception happened that has certain properties?
- What's the main driver for putting this in Zeek itself? I for sure see reasons here — if Zeek itself wants to use ZTest, it's more cumbersome to have to go through zkg to install it. I'm flagging this only because overall we've been trying to reduce the load of scripts shipped with Zeek. So if packages are the main intended user, I'm not as convinced about including ZTest now, because it would help the packaging system mature if we start to rely more on dependencies and diversify the purpose of packages. Then again, the unit testing topic clearly seems worthy of a framework, and it seems we'd likely put this somewhere like `base/frameworks/ztest`, right? That sounds good to me.

Basically, I think ZTest is a useful toolkit, but to me not yet a Zeek unit testing framework. And that's okay, ZTest doesn't have to have all answers right now. But we should make that clear in the docs and release notes.

**Comment by ckreibich** [comment]
@bbannier flagged that arguing for package dependencies won't hold up very well as long zeek/package-manager#106 isn't resolved. :smile:

**Comment by bbannier** [comment]
Since this appeared in the zeek-4.2 roadmap I'll take a stab at where this is currently stuck, but I think @ckreibich's [comment above](https://github.com/zeek/zeek/issues/1475#issuecomment-865276835) is already a great summary.

1. I arrived at this topic by making a careless comment about testing in Zeek scripts, specifically I was looking for at least an `assert` function (there could also be any of `assert_[false|eq|ne|lt|gt|le|ge]`). I believe one could make a general case for adding such builtin functions to Zeek proper, outside of a package as they would make test code more explicit (the current approach is to `print` values and then `btest-diff` the output), as well as also potentially being useful in production code to check invariants.
2. Testing frameworks typically provide a way to introduce scopes for isolated tests (also used to e.g., register tests or sometimes execute individual tests). ZTest provides functionality for that via `ZTest::test`, but since this function just schedules callbacks for execution in the same Zeek process, they are probably not isolated enough to test aspects involving e.g., global Zeek state. If this matters one would probably still end up writing separate BTest tests.
3. ZTest doesn't provide any solutions around mocking input, or tooling to hook into events or logs. I believe it should be possible to implement such functionality independently of ZTest, to a large extent probably in a package.

At this point to me only (1) seems realistic; (2) and (3) need more thinking, and it isn't even clear at this point that anybody is interested in a solution here.

**Comment by rsmmr** [comment]
I pulled out the idea of an `assert` into a discussion: https://github.com/zeek/zeek/discussions/1792

**Comment by bbannier** [comment]
Unassigning myself while we figure out what we want to provide.
